1657 resource file api

alvacouch opened the pull request:
@pkdash @hyi This is submitted for early review. It addresses one small part of issue #1657. It is built on top of PR #1672 so one should review that one first! It affects folders, files, and file management. 

The dynamics of `FileField` are so complex that we are at present ignoring any builtin `FileField` functions. Our code contains many statements that break that object, including variants of resource_file="string" that -- in effect -- corrupt the file path to the object. Moving forward, it would be nice if we could rely upon `FileField` for file management again. This PR is a first step in that direction. 

This PR implements a method for normalizing storage paths so that `IrodsStorage` functions remain functional after paths are moved. This has two main parts: 
* `ResourceFile.set_storage_path`: Sets a new storage path for a `FileField` bound to `IrodsStorage` or `IrodsStorage("federated")` so that the storage object remains functional. 
* `get_path`: a modified target to `FileField.upload_to` that handles whatever FileField hands to it and outputs a normalized path.  

**The point of this is that if whenever you want to move a file, you use `set_storage_path`, the `IrodsStorage` object will get a meaningful path and can continue to be used.**

As well, `ResourceFile.storage_path` will at any time regurgitate an iRODS file path suitable for use with iCommands. Sophisticated error handling rejects erroneous paths. Thus, the `FileField`s provide definitive information on the states of local or federated files, respectively.

alvacouch commented:
@hyi @pkdash Tight closure achieved, in the sense that 
1. Every call to set_storage_path leads to something that can be interpreted in IrodsStorage. 
2. This includes path syntax and existence of file. 
This will solve many problems if used consistently.

pkdash commented:
@alvacouch It seems pylint checks are failing.

mjstealey commented:
> pylint checks are failing

```
./hs_core/tests/api/native/test_resource_file.py:109:49: W291 trailing whitespace
./hs_core/tests/api/native/test_resource_file.py:111:28: W291 trailing whitespace
./hs_core/tests/api/native/test_resource_file.py:168:49: W291 trailing whitespace
./hs_core/tests/api/native/test_resource_file.py:170:28: W291 trailing whitespace
```

hyi commented on hs_core/models.py:
```diff
@@ -1508,41 +1512,205 @@ class Meta:
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
     fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
+    # TODO: why is size a CharField???
     fed_resource_file_size = models.CharField(max_length=15, null=True, blank=True)
```
@alvacouch Agreed although it works to use a string to store the file size, it makes more sense to have fed_resource_file_size a BigIntegerField rather than a CharField. Feel free to correct it as needed.

hyi replied:
@alvacouch If we refactor to make fed_resource_file as single ground truth for federated zone files, we don't need this fed_resoruce_file_size field. Yes, I agree the size() property is not accurate as of now and need to add fed_resource_file.size as well.

hyi commented on hs_core/models.py:
```diff
@@ -1508,41 +1512,205 @@ class Meta:
+        if self.content_object.resource_federation_path:  # false if None or empty
+            return self.fed_resource_file.name
+        else:
+            return self.resource_file.name
```
Even if ```resource_federation_path``` is not empty which indicates the resource is stored in a federated zone, ```fed_resource_file``` would be very likely be empty for majority of cases when a user selects a file from a federated zone using iRODS browser. In this case, ```fed_resource_file_name_or_path``` and ```fed_resource_file_size``` store the relative path in a federated zone where the resource is created, and there is no file upload action, hence ```fed_resource_file``` is empty. ```fed_resource_file``` is only set when a file from local disk is uploaded and added to an existing resource in a federated zone, or when the resource is initially created, the user selected a file from a federated zone using iRODS browser, then select another file from local disk, in which case, the resource will be created in the iRODS federated zone and fed_resource_file is set to associate with the file uploaded from local disk. Hence another conditional check is needed when ```resource_federation_path``` is true to see whether fed_resource_file_name_or_path is empty or not, and also checks whether fed_resource_file is empty or not, since they represent different cases and may be empty.

alvacouch replied:
This is incredibly -- and in my opinion unnecessarily -- confusing. Are you saying that when one uses irods browser to load a file, that it isn't even stored in the resource, but rather, in a user home directory, it is not copied into a federated hydroproxy zone? That is really not recommended. This creates states of a resource that will be difficult to deal with.

hyi replied:
@alvacouch In other words, only a reference point is needed to a file stored in a federated zone selected using iRODS browser, no real file object is stored in ResourceFile object. Does it make sense to you?

alvacouch replied:
I would vastly prefer to use fed_resource_file for all federation, but this would mean that all files would have to have the same federation path prefix.

hyi replied:
@alvacouch ```fed_resource_file``` is defined as a FileField, which is a file-upload field in Django that handles file upload to a storage. In our federation case, there is no file to upload from if the file comes from a federated zone, so this ```fed_resource_file``` FileField cannot be used. This is a dilemma I faced when I initially implemented this. In short, we cannot use FileField for files directly coming from a federated zone.

alvacouch replied:
@hyi There is no particular contradiction in using a FileField for things received from an iRODS zone. If the federation paths are the same, there is no need to create another object. We won't be using file upload, but the fed_resource_file object -- as ground truth for where the file lives -- is fine to use. Thus, a variable is eliminated, along with a lot of logic. 
 
If -- on the other hand -- the federation paths are different for the two kinds of objects -- I have great concern that this creates a management burden.

hyi replied:
@alvacouch After giving this some thought, I think it might just work to use fed_resource_file as a single ground truth for files that directly come from a federated zone without requiring any file upload. At the time when I initially implemented this, I thought this file upload field could not be directly used. Now that I am looking into the code, I think what you proposed might just work. This will simplify things for sure, although data migration may have issues. Do you think you can tackle this in this PR? If not, you can create an issue for this refactoring and assign it to you if you'd like to work on it or assign it to me otherwise.

hyi commented on hs_core/models.py:
```diff
@@ -1508,41 +1512,205 @@ class Meta:
+        Path can be absolute or relative.
+
+            * absolute paths contain full irods path to local or federated object.
+            * relative paths start with anything else and ican start with optional folder
```
typo for ```ican```

hyi commented on hs_core/models.py:
```diff
@@ -1508,41 +1512,205 @@ class Meta:
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(file_path)
+            relpath = relpath[plen:]  # strip local prefix, omit /
+
```
@alvacouch up to this point, relpath has already been check to make sure it exists in iRODS storage, so the following check perhaps should be put into an else condition so if the path has already been checked to have existed in the code above, it does not have to go through the exist check again in the code below?

hyi commented on hs_core/models.py:
```diff
@@ -1508,41 +1512,205 @@ class Meta:
+        from hs_core.views.utils import remove_folder
+        _path_is_allowed(folder)
+        # TODO: move code from location used below to here
+        remove_folder(user, resource.short_id, folder)
```
@alvacouch I don't understand why these class methods need to be put into a ResourceFile class. ```create_folder()``` and ```remove_folder``` is resource level operations. One resource can contain multiple ResourceFile instances with each one associated with one file included in the resource, so I don't think these class methods belong here.

alvacouch replied:
These methods interact with files in an intimate way.

hyi replied:
I know logically it sounds more reasonable to put file operations in ResourceFile class, however, our ResourceFile class definition has a many-to-one foreign key relationship with BaseResource class, so one resource contains multiple ResourceFile objects with each ResourceFile object corresponding to each single resource file. ```create_folder``` and ```remove_folder``` is resource level operations since it is not related to a single contained resource file instance. So I think folder-level operations belong to BaseResource class methods or as external utility methods.

hyi replied:
@alvacouch Although it is always nice to isolate iRODS interaction to one class, I don't think putting these folder related operations in ResourceFile class is appropriate since semantically it is not correct as ResourceFile as currently defined only corresponds to one single resource file object and these folder operations are beyond one single file but multiple files and folders under a resource. So I think BaseResource class is a better place to put them in. Or make it better encapsulated, perhaps add a ResourceFiles OneToOneField to BaseResource class and make each ResourceFile a ForeignKey many-to-one relationship to ResourceFiles. However, I am afraid we would have troubles in data migration if we make a change like that.

hyi commented:
@alvacouch I have finished the first round of review and left some comments.

pkdash commented on hs_core/tests/api/native/test_resource_file.py:
```diff
@@ -204,11 +203,14 @@ def test_federated_root_path_logic(self):
-        resfile.resource_file = None
-        resfile.save()
 
+        print("storage path is " + str(resfile.storage_path))
```
Not sure if we should be using print statements inside unit tests.

alvacouch replied:
I'll get rid of them. I was using it to debug. A simple oversight.

pkdash commented:
@alvacouch Is this ready for another round of review? If so let @hyi know.

alvacouch commented:
@pkdash @hyi Not ready for another round of review yet. Need to flesh out details.

alvacouch commented:
@hyi @pkdash There are many merge conflicts to fix and the migration to finish, but this version is stable in the sense that the basic patterns are not going to change. Could you take a look? There will be no surprises in the merge conflicts; the same pattern will be used everywhere.  

This is the first in a series of "maintainability changes" in the sense that I really didn't change anything except the maintainability of the code. I did this by enforcing a single point of truth for each file operation. This -- in turn -- makes a lot of things possible that were unreliable before. E.g., ResourceFile.resource_file.open works!

I am pretty much giving up on fixing the irods problems I've found (additional and missing files) in a migration. I tried a few approaches and it was painful.  I have a migration nearly ready that simply modifies filenames appropriately without interacting with iRODS. I will need to follow up with application level admin code to do the iRODS/Django cleanup. At that level, it will be a lot easier to write. 

What makes cleanup incredibly difficult during a migration is that **none of the model methods are available during a migration.** I had to copy everything I needed into the migration code. This makes sophisticated things -- like cleaning up unreferenced files in iRODS -- really difficult to accomplish within a migration.

alvacouch commented:
@pkdash @hyi Simplified migration is now posted. This will have to do for the migration. I will have to do other things to clean up excess files and dangling pointers.

alvacouch commented:
...Fixing bugs in migration using current Django database.

alvacouch commented:
@hyi @pkdash Migration now corrects all known errors. Entire repository comes up clean. 

Notes: 
1. It seems that no resource uses `ResourceFile.fed_resource_file` at all. All federated resources seem to use `ResourceFile.fed_resource_file_name_or_path`. I moved these values to Resource.fed_resource_file and fixed that file so it works and provides its basic functions. 

2. I set the file path to None whenever it was `""` or `"None"` (the string). Some resources had this. 

3. There are many more cases in the migration than actually exist in the resources. Each case evolved as I tracked down errors. Putting `"None"` rather than `None` in for a filename was misinterpreted as several different errors before I found the truth of it. 

This migration code is the beginning of an integrity checker for files that can be run nightly or weekly to detect programming errors.

hyi commented:
@alvacouch I have not got a chance to look into your PR yet, but your migration code for file integrity check sounds useful and much needed. I want to answer your question on ResourceFile.fed_resource_file not being used currently. The fed_resource_file field is specifically created to handle the case in which the file was originally from local disk being uploaded, but instead of being stored in hydroshare zone, it is stored in federated iRODS zone such as hydroshare user zone. This special case only happens when on the resource creation page, the user selected files using iRODS browser from a federated zone such as hydroshare user zone which warrants us to store the to-be-created resource in the hydroshare user zone rather than hydroshare zone, then in addition to those files, the user also uploaded files from local disk by selecting the file browsing button. Since we allow users to select files from local disk and from iRODS before creating a resource, this case could happen, and the fed_resource_file field is created specially to handle this case. You don't see it being used currently on iRODS side since nobody has not done this yet, but could do it. I was not comfortable with this design/implementation at the time to handle this special case, but could not think of a better approach at the time and did not want to be held off by that. If you have better idea to handle this, let me know or you can improve/change it in this PR.

alvacouch commented:
@hyi I don't think we need a separate File or path object based upon the source of the data. We've already decided separately that we're always going to make a copy of any user zone files and never append them "in place" (to avoid a lot of issues). 

So the only distinguishing mark between different kinds of files is whether they're federated or not. So I've modified the code so that those are the two use cases. 

If we need to remember the source, we can always record that source as metadata...!

(I noticed also that some of your code -- ostensibly intended for copying from a federated user zone to a federated HydroShare zone -- could easily be made portable so it also works for local user zone to local HydroShare zone copies. So I did so.)

hyi commented:
@alvacouch Can you test in your local development environment with your new code that removed fed_resource_file field that it can handle the case when a user selects a file from local disk, then selects a file from his or her federated user zone to see if there is any issues? I remember it had issues when I implemented federation functionality which is why I added fed_resource_file field. So it'd be good if you can test it out to make sure your updated code can handle this case without issues.

alvacouch commented:
@hyi @pkdash : will test in federated mode again. Tests passed last time I checked, before today's push to resolve conflicts. Confidence is high, though. 

I arranged for all `FileField` objects -- both `resource_file` and `fed_resource_file` -- to contain full pathnames. Several helper functions help one manipulate these: 

* `ResourceFile.storage_path` is the full path to the file, either federated or not. 
* `ResourceFile.short_path` is the relative path not including data/contents. 

Thus, for a file `f`, `f.fed_resource_file_name_or_path == os.path.join('data/contents', f.short_path)`. I made this substitution in your scripts, and they run without changes.

pkdash commented:
@alvacouch Let me know when it is ready for review.

hyi commented:
@alvacouch It appears there are still conflicts to be resolved with develop branch for this PR. Also, I don't think I have written the test case to capture the scenario where fed_resource_file is intended to be used, i.e., one selects a file from a federated user zone, then select a file from local disk to upload, for creating a resource, an oversight of course. Now that you are at it, it'd be great to write this test case to make sure it passes with your updates. Let me know if you need help from me in adding this test case.

alvacouch commented:
@pkdash @hyi I am resolving further conflicts now. Done by end of day. 

@hyi The test case that you wrote for fed_resource_file_name_or_path is now a test case for fed_resource_file...!

hyi commented:
@alvacouch It'd be great if existing test cases already cover the specific, not common, scenario I described above. I'll keep this in mind when I review this PR. Agreed the logic will be much simplified if we can remove fed_resource_file field. Thanks for doing this.

alvacouch commented:
@pkdash @hyi  I will continue testing this, but I think it's ready for review. All merge conflicts are resolved.

alvacouch commented:
@pkdash Can't figure out how to reference the latest django_irods SHA..... can you help?

pkdash commented:
@alvacouch to use the latest django_irods:

```cd django_irods```
```git checkout origin master```
```cd ..```
Then you may have to add the submodule
```git add django_irods```

Then commit and push

mjstealey commented:
> cd django_irods
> git checkout origin master
> cd ..

May want to put a fetch and a pull in there to ensure you have the latest pointers.

```
cd django_irods\
git fetch origin -p
git checkout origin master
git pull
cd ../
```

And then the add / commit piece

hyi commented:
@alvacouch Looks like you did not update django_irods submodule SHA reference in your branch properly, hence it still shows the django_irods changes which have already been merged into master. I also checked your branch and found django_irods submodule did not refer to the master branch of django_irods repo.

hyi commented on hs_collection_resource/utils.py:
```diff
@@ -59,6 +59,7 @@ def update_collection_list_csv(collection_obj):
         # remove all files in bag
+        # TODO: I do not understand precisely why this is necessary
         for f in collection_obj.files.all():
             delete_resource_file_only(collection_obj, f)
```
@zhiyuli Can you comment on why this is necessary to address @alvacouch 's TODO comment above?

pkdash replied:
It seems it is trying to delete the only possible file (which  a csv file) for the collection resource before it adds an updated csv file to the resource.

hyi replied:
@pkdash Yes, I believe you are right.

alvacouch replied:
Code comment updated.

hyi commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -29,55 +27,47 @@ def delete_bag(resource):
-        istorage.delete(res_path)
+    # TODO: Is this correct? Is it really necessary? It seems this deletes all content!
+    if istorage.exists(resource.root_path):
+        istorage.delete(resource.root_path)
```
@alvacouch delete_bag() function does delete the whole resource content, since it is called by delete resource action. Although resource files are deleted as part of delete() call, the other files created as part of content bag still exist such as manifest files, metadata xml and resourcemap xml files, etc., which will be deleted as part of this delete_bag() call, which is what is accomplished here.

alvacouch replied:
This needs to be renamed. I would assume that it deletes the bag and not the resource.

alvacouch replied:
This has been renamed to delete_files_and_bag

hyi commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -29,55 +27,47 @@ def delete_bag(resource):
     # delete the bags table
+    # TODO: Why were extra bags created in the first place?
     for bag in resource.bags.all():
         bag.delete()
```
@alvacouch This Bags objects on resources don't hold anything meaningful at this point. It exists as part of HydroShare legacy before we moved bag creation to on-demand bag creation on iRODS side. It can be removed in my opinion. I remembered I intended to remove it, but there were some issues probably migration-related, so I did not really take it on at the time.

hyi commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -29,55 +27,47 @@ def delete_bag(resource):
 
 
-def create_bag_files(resource, fed_zone_home_path=''):
+def create_bag_files(resource):
```
@alvacouch This PR needs to be merged before the next deployment now that create_bag_files() signature has been changed already in django_irods where fed_zone_home_path parameter is not passed in any more. Otherwise, bag files will not be able to be created for resources in a federated zone. Just a note to keep in mind. Keeping @mjstealey informed to make sure this PR needs to be merged before the next deployment.

alvacouch replied:
@hyi @mjstealey If that is too confusing I could simply accept and ignore the file path to make the deploy work seamlessly. However -- and this is much more important -- the new semantics for files -- including no overwriting of existing files -- caused lots of tests to break. I have fixed those tests in this PR only.

alvacouch commented:
@pkdash There is considerable confusion in your hs_composite_resource code that I am attempting to clear up in this PR. In your code, you pass three different kinds of paths: 

* {folder}/{filename} 
* data/contents/{folder}/{filename}
* {irods-path}/data/contents/{folder}/{filename} 

After my revision, the second one is considered illegal; there is no purpose to it. I have eliminated it to the extent of my abilities. However, the remaining two remain. To my mind, it makes sense to only pass the first, because {resource}.file_path contains the rest, so the full information is redundant. Do you agree? 

(I could also support both, but would rather not. Seems redundant.)

pkdash commented:
@alvacouch If the first one is the correct way to do it, then we should only do that. I have yet to check where these 3 different paths are being passed. If you can do the cleanup as part of your PR that would be great. Thanks.

pkdash commented:
I plan to review this PR later today.

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -263,7 +267,7 @@ def create_resource(
         edit_users=None, view_users=None, edit_groups=None, view_groups=None,
         keywords=(), metadata=None, extra_metadata=None,
-        files=(), fed_res_file_names='', fed_res_path='', fed_copy_or_move=None,
+        files=(), source_names=[], fed_res_path='', move=False,
```
@alvacouch According to the docstring, this source_names is a string, so the default should be ```''``` rather than ```[]```

hyi replied:
@alvacouch OK, I see source_names used in the code is actually a list, in which case, the docstring should be updated to reflect this.

alvacouch replied:
Updated docstrings to clarify, and eliminated all use of source_names as a string. There was one remaining place where it **was** a string.

pkdash commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -90,24 +80,23 @@ def create_bag_files(resource, fed_zone_home_path=''):
-        to_file_name = '{fed_zone_home_path}/{rel_path}'.format(
-                            fed_zone_home_path=fed_zone_home_path,
-                            rel_path=to_file_name)
+        out.write(resource.metadata.get_xml())
```
Unless the function ```get_xml()``` has been changed in this PR, it will only generate resource level metadata and not file level metadata.

alvacouch replied:
Yes, it has not been modified and this is resource-level metadata. 
Is that an error?

pkdash replied:
In that case keep the original statement:
``` out.write(resource.get_metadata_xml()) ```

Resource and file level metadata is written to the same xml file.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -188,6 +188,7 @@ def get_federated_zone_home_path(filepath):
 
 
+# TODO: this is replaced with ResourceFile.size, for both federated and unfederated resources
 def get_fed_zone_file_size(fname):
```
Should this function be deleted then?

alvacouch replied:
Yes, but in next PR.

alvacouch replied:
Deleted. No more references present.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -844,105 +831,65 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
     :return: The identifier of the ResourceFile added.
     """
     if f:
+        # does not seem to be needed
```
Remove these commented code?

alvacouch replied:
It's in the code now and I don't know why. 
Would prefer to leave it in as documentation in case things break. 
Then I'll know what to try........

alvacouch replied:
Verified that deletion is safe. Deleted.

pkdash commented on hs_core/models.py:
```diff
@@ -1429,14 +1431,19 @@ def supports_folders(self):
         """ returns whether folder operations are supported. Computed for polymorphic types"""
         return False
 
+    # TODO: Why isn't this a regular method? Why does it need to be a class method?
```
Agreed. This should be implemented as property.

alvacouch replied:
Will make this change in a further PR.

pkdash commented on hs_core/models.py:
```diff
@@ -1447,55 +1454,47 @@ def scimeta_url(cls, resource_id):
         scimeta_url = reverse('rest_download', kwargs={'path': scimeta_path})
         return scimeta_url
 
+    # TODO: there are too many ways to get to the resourcemap.
```
This should be a property too.

pkdash commented on hs_core/models.py:
```diff
@@ -1642,28 +1641,46 @@ def can_be_public_or_discoverable(self):
 
         return False
 
+    # TODO: these should be instance methods, not class methods.
```
This can be still called from an instance right?

alvacouch replied:
Yes, but they act on instances, so they should not be class methods.

pkdash commented on hs_core/testing.py:
```diff
@@ -180,14 +168,19 @@ def resource_file_oprs(self):
-            add_resource_files(res.short_id, fed_res_file_names=[fed_test_file1_full_path],
-                               fed_copy_or_move='copy',
-                               fed_zone_home_path=res.resource_federation_path)
+            # TODO: why isn't this a method of resource?
```
Agreed.  This should be implemented a function of the BaseResource class. May be in a follow up PR one of us can do some additional internal API rework.

pkdash replied:
Sounds good.

pkdash commented on hs_core/tests/api/native/test_bagit.py:
```diff
@@ -50,6 +50,7 @@ def test_create_bag(self):
     def test_create_bag_files(self):
         # this is the api call we are testing
         irods_storage_obj = hs_bagit.create_bag_files(self.test_res)
+        print(type(irods_storage_obj))
```
Not sure if we should have ```print``` statements?

pkdash commented on hs_core/tests/api/native/test_core_metadata.py:
```diff
@@ -862,9 +878,17 @@ def test_format_element_auto_deletion(self):
         fmt_element = res.metadata.formats.all().first()
         self.assertEqual(fmt_element.value, format_CSV)
 
+        print("before delete resource file") 
```
Not sure if we need/should have these print statements.

alvacouch replied:
They're debugging printouts. Will delete all of them.

alvacouch commented:
@pkdash 
A general comment. It is a general principle that one should not "require" the world outside a module to understand things inside a module. In the case of file handling, the file handling system knows where files should be stored, and all it needs is the subdirectory. Explicitly forcing someone outside the module to list the full path is not just redundant, but also brittle, because mistakes in that path will cause unknown behaviors. 

This is true system wide, however, and has been a lot of work to fix.

alvacouch commented:
See PR #1898 for the part of this that depends upon function of django_irods. That will have to be merged here as well.

alvacouch commented:
@hyi @pkdash The other PR gives me some flexibility now. I am at a crossroads now in the sense that I can: 

1. Finish this PR at the cost of leaving some things awkward. What is still awkward is what I pointed out above: the API functions for manipulating resource files still take three distinct forms of pathnames as arguments. This really needs to be fixed. It's incredibly confusing to a programmer. I know why it is that way, but the reason is obsolete. The functions for ResourceFiles all act on {prefix}/data/contents/* and thus those routines -- specifically -- can be simplified. (This excepts bags, but bags have their own function bag_path that locates them unambiguously).  

2. Clean up more, but at the same time, risk needing to rebase **again**. I've already rebased twice!

pkdash commented:
@alvacouch I suggest to go for option#1 above. The other changes (multiple formats) should be fixed in a follow up PR.

pkdash commented on hs_core/tests/api/native/test_resource_file.py:
```diff
@@ -0,0 +1,330 @@
+
+        self.assertTrue(resfile.path_is_acceptable(shortpath, test_exists=False))
+
+        # non-existent files should raise error
```
Not seeing it raising any error.

alvacouch replied:
Fixed. Was incomplete test.

pkdash commented on hs_core/tests/api/native/test_resource_file.py:
```diff
@@ -0,0 +1,330 @@
+
+        self.assertTrue(resfile.path_is_acceptable(shortpath, test_exists=False))
+
+        # non-existent files should raise error
```
Not seeing error being raised.

alvacouch replied:
Fixed in latest. A test I never finished.

pkdash commented on hs_core/tests/api/native/test_resource_file.py:
```diff
@@ -0,0 +1,330 @@
+
+        # create a file
+        self.test_file_name1 = 'file1.txt'
+        # self.test_file_name2 = 'file2.txt'
```
I assume these commented code statements here in other places in this file can be removed.

alvacouch replied:
Done.

pkdash commented on hs_core/models.py:
```diff
@@ -1750,70 +1767,494 @@ class Meta:
+    return get_resource_path(instance.resource, filename, folder)
+
+
+def get_resource_path(resource, filename, folder=None):
```
Should this function be called ```get_resource_file_path()``` since it is computing path for a FileField?

pkdash replied:
Should this be an instance function of the BaseResource?

alvacouch replied:
Changed name; put in a TODO for moving it.

pkdash commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+                                                                            folder))
+
+    @classmethod
+    def list(cls, resource, folder):
```
Wondering why this is not an instance function of the BaseResource?

hyi replied:
@alvacouch This sounds to me like an instance method of the BaseResource as well.

alvacouch replied:
Changed name. Put in TODO for moving it.

pkdash commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+                resource_file__startswith=folder)
+
+    @classmethod
+    def create_folder(cls, resource, folder):
```
Comment as above.

alvacouch replied:
Added TODO for moving this in next PR.

pkdash commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+        create_folder(resource.short_id, os.path.join('data', 'contents', folder))
+
+    @classmethod
+    def remove_folder(cls, resource, folder, user):
```
Same as above.

alvacouch replied:
Added TODO for moving this in next PR.

pkdash commented:
@alvacouch I have finished reviewing this PR and have left few comments.

hyi commented:
@alvacouch I have not been able to finish reviewing this PR yet since I have been sidetracked to work on other items. In the process of looking into an issue raised by @gantian127 with netCDF large file resource copying (not sure why that issue only happens for that particular published resource, but not with other resources stored in the federated user zone yet, still in the process of investigating), I found fed_resource_file field in ResourceFile class holds the full path of the netCDF header file in the format of /irods_federated_zone/home/localHydroProxy/resource_id/data/contents/header.txt, which is inconsistent with other fields in ResourceFile class. Feel free to fix that now that you are at it. I can help with migration of existing resources in federated user zone if needed for this.

hyi commented on hs_app_timeseries/receivers.py:
```diff
@@ -35,9 +35,12 @@ def pre_add_files_to_resource_handler(sender, **kwargs):
-    if files or fed_res_fnames:
+    if __debug__:
+        assert(isinstance(source_names, list))
+
```
@alvacouch Make sure you remove these debug asserts before merging this PR.

alvacouch replied:
@hyi @pkdash I would prefer to leave ones like this in temporarily, because in the next PR, I will use it as a sanity check that I am doing things correctly. This will prevent me from making common errors during porting. The __debug__ clauses are not invoked during production. 

What I am doing is difficult enough that I am triple-checking my code with __debug__ clauses that will catch errors I could make. I think anyone in my position would.

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -2,7 +2,7 @@
 import shutil
 import logging
-import string
+# import string
```
@alvacouch The commented ```import string``` should be removed if it is not used, right?

alvacouch replied:
Done

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -305,25 +309,26 @@ def create_resource(
-                               used to create the resource in the federated zone, default is empty
-                               string
+    :param source_names: an array of file names separated by comma from a federated zone to be
+         used to create the resource in the federated zone, default is empty array
```
@alvacouch The docstring for source_names does not sound right to me. If it is a list, there is no need for ```file names separated by comma``` but instead all file names are stored as a list, right?

alvacouch replied:
This is a mistake. Will fix. It used to be a string and I changed its format.

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -354,11 +359,13 @@ def create_resource(
-            fed_zone_home_path = utils.get_federated_zone_home_path(fed_res_file_names[0])
+
+        # TODO: this option should be eliminated. Require explicit zone path
+        elif len(source_names) > 0:
```
@alvacouch On your TODO comment, I believe this option cannot be eliminated since the workflow is to allow users to browse into any iRODS zone, then we detect whether it comes from a federated zone, and if it is, then we get the federated zone path based on ```source_names``` which are the iRODS absolute logic path to selected iRODS files, then pass source_names to create_resource() function which is used to form the resource_federation_path.

alvacouch replied:
This behavior is problematic because the user can pass polymorphic zone paths and there is no check that they are in the same zone. An alternative is to explicitly check that the zones are all the same.

hyi replied:
@alvacouch I understand your point, but at this point, iRODS browser only allow users to browse into the same zone, if users want to browse into another zone, they have to login again, which clears up the previous iRODS file selection and restarts another iRODS file selection. In short, we don't allow users to browse into multiple different zones at the same time as is implemented currently. So you can update your TO DO comments to reflect this.

alvacouch replied:
TODO updated.

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -438,8 +444,8 @@ def create_empty_resource(pk, user, action='version'):
-        the empty new resource that is created as an initial new version or copy for the original
-        resource which is then further populated with metadata and content in a subsequent step
+        the empty new resource that is created as an initial new version for the original resource
+        which is then further populated with metadata and content in a subsequent step
```
@alvacouch This function is used by both creating new version of a resource and creating a new copy of a resource, so the original comment should be kept or made better but not remove ```copy``` since it is important to note this function is used by both functionalities.

alvacouch replied:
Fixed.

hyi commented on hs_core/hydroshare/resource.py:
```diff
@@ -577,57 +583,38 @@ def add_resource_files(pk, *files, **kwargs):
+            ret.append(utils.add_file_to_resource(resource, None,
+                                                  folder=folder,
+                                                  source_name=ifname,
+                                                  move=move))
```
@alvacouch Has the string option eliminated? If not, this for statement will fail to do what it is supposed to do for string option, right? It is most likely no calling function passes a string to source_names. But if there is, it should be changed to list so that it can work here.

alvacouch replied:
It needs to be changed to a list. Will do.

hyi replied:
@alvacouch The TO DO comment is a bit confusing here. Has the string option been eliminated? Looks like it has. If so, the TO DO comment needs to be updated. Also, a minor comment is that ```if len(source_names) > 0:``` can be changed to ```if source_names:``` to be simpler, but both will work.

alvacouch replied:
TODO no longer needed and deleted.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -249,20 +258,9 @@ def get_file_from_irods(res_file):
-        file_name = os.path.basename(res_file.fed_resource_file_name_or_path)
+    istorage = res.get_irods_storage()
+    res_file_path = res_file.storage_path
+    file_name = os.path.basename(res_file_path)
```
@alvacouch @pkdash I am afraid there might be a chicken-and-egg issue here (although I am not sure since this might not be an issue any more with the changes in this PR). Specifically, if a specific resource type is listening to pre_create_resource signal before the resource is created to extract metadata, the storage_path is most likely not set at this point since the resource is not created yet, but set on ```res_file.fed_resource_file_name_or_path```? Worth checking to make sure.

hyi replied:
@alvacouch @pkdash Sorry, not ```storage_path```, but ```res.get_irods_storage()``` is what I meant above.

hyi replied:
@alvacouch @pkdash OK. I take my comments back. I think it is fine at least in production. You can disregard my comment above. But I will not delete my comment for reference just in case.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -398,6 +385,7 @@ def replicate_resource_bag_to_user_zone(user, res_id):
     # needs to check whether res_id collection exists before getting/setting AVU on it to
     # accommodate the case where the very same resource gets deleted by another request when
     # it is getting downloaded
+    # TODO: why would we want to do anything at all if the resource does not exist???
```
@alvacouch I agree, you can change to exit out without continuing if the resource does not exist, although this is just a precaution since it is called from a resource landing page and the resource should exist at this point. But in the future when this is exposed to REST API, there could be possibility the resource does not exist.

alvacouch replied:
Changed to raise ValidationError if resource doesn't exist.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -428,58 +416,38 @@ def copy_resource_files_and_AVUs(src_res_id, dest_res_id, set_to_private=False):
     tgt_res = get_resource_by_shortkey(dest_res_id)
+
+    # This makes the assumption that the destination is in the same exact zone.
+    # Also, bags and similar attached files are not copied.
```
@alvacouch Yes, copying action intends to copy a resource to the same zone, which is a valid assumption I think. And yes, bags are intentionally not copied over since it is only generated on-demand when users want to download the bag.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -428,58 +416,38 @@ def copy_resource_files_and_AVUs(src_res_id, dest_res_id, set_to_private=False):
-        dest_coll += '/'
-    else:
-        dest_coll = dest_res_id + '/'
+    dest_coll = os.path.join(tgt_res.root_path, 'data')
```
@alvacouch dest_coll should not include ```data``` sub-directory path since it will be copied over from src_coll. Otherwise, a sub-directory ```data``` will be created under ```data```.

hyi replied:
@alvacouch I did notice there is a trailing ```/``` included in the original code, so ```tgt_res.root_path``` may need to be concatenated with a trailing ```/``` to have the copy work as expected on iRODS side, but I am not 100% sure about it.

alvacouch replied:
I also changed the src_coll so that this copy works properly.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -428,58 +416,38 @@ def copy_resource_files_and_AVUs(src_res_id, dest_res_id, set_to_private=False):
+    dest_coll = os.path.join(tgt_res.root_path, 'data')
 
+    # TODO: This copies everything but bags, including unmapped junk files(!)
     istorage.copyFiles(src_coll, dest_coll)
```
@alvacouch Addressing your TODO comment, this is intended, we should not copy bags and manifest files since the user may changes things around after the copy action, then the bags and manifest files need to be regenerated anyway, so there is no point to copy the bag and manifest files over.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -544,18 +520,29 @@ def resource_modified(resource, by_user=None, overwrite_bag=True):
+    to indicate that the resource is modified for on-demand bagging.
+
+    set bag_modified (AVU) to 'true' for the modified resource in iRODS to indicate
+    the resource is modified for on-demand bagging.
```
@alvacouch Appears the comment sections above are duplicated.

alvacouch replied:
Duplicate removed.

hyi commented on hs_core/hydroshare/utils.py:
```diff
@@ -844,105 +838,65 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
-            # delete the file added if there is any exception
-            ret.delete()
             # raise the exception for the calling function to inform the error on the page interface
             raise SessionException(ex.exitcode, ex.stdout, ex.stderr)
```
@alvacouch I think it is safer to include that ```ret.delete()``` if there is an exception since the exception is caused by iRODS at which point, a record may have already been created in Django database since Django DB operation and iRODS file operation are not atomic, so we don't want to leave any "gabage" in django database if there is an iRODS exception.

alvacouch replied:
Because delete now cascades to files, it could fail due to a partly created ResourceFile. I enclosed it in a try block.

hyi commented on hs_core/models.py:
```diff
@@ -1429,14 +1431,19 @@ def supports_folders(self):
 
+    # TODO: Why isn't this a regular method? Why does it need to be a class method?
+    # It would seem to me that one only creates a bag after a resource has been created,
+    # so that this would be an instance method....
```
@alvacouch Agreed, this sounds to me like an instance method rather than a class method as well.

alvacouch replied:
Will make this change in a further PR.

hyi commented on hs_core/models.py:
```diff
@@ -1447,55 +1454,46 @@ def scimeta_url(cls, resource_id):
+    # TODO: this is inaccurate; resourcemap.xml != systemmetadata.xml
     @classmethod
     def sysmeta_path(cls, resource_id):
         return "{resource_id}/data/resourcemap.xml".format(resource_id=resource_id)
```
@alvacouch Agreed. Why don't we call the method ```resourcemap_path``` and this sounds like an instance method rather than class method too.

alvacouch replied:
We actually need a sysmeta_path because of DataOne requirements. The routine's definition is incorrect.

pkdash replied:
We have a REST endpoint for system metadata. I am not sure if it provides the correct data to the caller though.

alvacouch replied:
@pkdash The REST endpoint for system metadata returns resource metadata. This is a bit of an embarrassment. I will get to fixing this as part of integrating with DataOne.

pkdash replied:
@alvacouch If you referring to the ```sysmeta_path()``` then that is not part of the REST endpoint for getting system metadata.  See the following view function that serves the system metadata endpoint:
https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/resource_rest_api.py#L531

hyi commented on hs_core/models.py:
```diff
@@ -1447,55 +1454,46 @@ def scimeta_url(cls, resource_id):
                 fl.logical_file.metadata.delete()
-
+            # COUCH: delete of file objects now cascades.
             fl.delete()
```
@alvacouch Are you sure ```fl.delete()``` would delete file with only fed_resource_file_name_or_path set on iRODS side? In other words, will istorage.delete() be explicitly called in delete() method of ResourceFile class?

alvacouch replied:
See new definition of ResourceFile.delete which is not the default.

alvacouch replied:
@hyi @pkdash However, I just realized that there are several specific cases that -- due to the absence of a logical file abstraction in the original code -- don't have the knowledge to remove files from iRODS that are implied by the ResourceFile record and not explicitly referenced!  Specifically, geographic feature resource and similar cases. Oops.   

Neither the old code nor the new code fixes this. It's a file leak. Will consider how to fix it. Probably through a type-specific signal.

hyi replied:
@alvacouch Some files such as resourcemap.xml, resourcemetadata.xml, netCDF header file, raster vrt file, etc., are not stored in ResourceFile class, but present in iRODS, so ResourceFile delete() alone is not enough to delete all files for a resource. delete_bag() routine (I believe you renamed it in this PR) which is called by resource deletion action actually deletes the whole resource collection on iRODS side so that all these files that are not represented in ResourceFile class are deleted as a result of resource deletion. So the current implementation does not have this file leak (if I understand it correctly) issue and may be sufficient for our needs.

alvacouch replied:
@hyi I am not talking about /data/*.xml. I am referring to shapefiles that don't seem to have records in ResourceFile. I have to check whether this is true or not, but the existing code implies that it is. If it's not true, a particular path through the code is never taken. That is also important to know....

hyi commented on hs_core/models.py:
```diff
@@ -1642,28 +1640,46 @@ def can_be_public_or_discoverable(self):
+    # These decisions are used when operating upon an instance.
+    # That requires less overhead than using a class method.
     @classmethod
     def get_supported_upload_file_types(cls):
```
@alvacouch @pkdash I think these methods are called before a resource is created but after a resource type is selected from drop down list for use by file browser. So it sounds to me class method is accurate rather than instance method.

pkdash replied:
It seems the current use of this method is always like a class method (invoked from a class and not from an instance). Since this method is not dependent on a instance it seems it should be a class method.

alvacouch replied:
OK. Changed comment.

hyi commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
-        prefix = os.path.join(instance.content_object.short_id, 'data', 'contents')
+    if not folder:
+        folder = instance.file_folder
+    return get_resource_path(instance.resource, filename, folder)
```
@alvacouch Should this get_path() be combined with get_resource_path() below? They sound to me like doing the same thing. Also, I think get_resource_file_path() as the method name is more accurate.

alvacouch replied:
They have different arguments and are called in different contexts. 
* get_path is called during file upload and its arguments are predetermined. I cannot change them. 
* get_resource_path is for use outside file upload with more arguments. 
Will consider changing the name.

alvacouch replied:
Changed name of second one.

hyi commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+                                         null=True, blank=True, storage=FedStorage())
+
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
+    # fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
```
@alvacouch I am wondering whether it is sufficient to replace this fed_resource_file_name_or_path field with a property storage_path to handle this case. This field represents the third scenario for ResourceFile where both resource_file and fed_resource_file fields are empty and this field indicates the relative iRODS path for a file stored in a federated zone. Having it as a field in parallel with resource_file and fed_resource_file makes it queryable in the database table and we can even update the field in database for migrations or correct mistakes if needed. Replacing it with a property would not allow us to do these things, right? So I am not convinced it is a good idea to remove this database field.

alvacouch replied:
We could just as easily put a flag into the ResourceFile object that allows this query. Why do you want to query in this way?

hyi commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
+    # fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
+    # DEPRECATED: use native size routine
+    # fed_resource_file_size = models.CharField(max_length=15, null=True, blank=True)
```
@alvacouch This field goes together with fed_resource_file_name_or_path field. Also, removing these two fields requires data migration, but at this point I am not convinced we should remove them.

hyi replied:
@alvacouch After reviewing a bit more, it looks like you have tried to use fed_resource_file to cover this third scenario I mentioned above. This may be sufficient and definitely simplify things a lot. I will have to test it out to make sure each scenario works as expected, though. It'd be great if we can deploy this to a VM where local iRODS containers are used so we can test federation. @mjstealey Can we make a VM self-deployment to use local iRODS containers so that we can test federations? It'd be the best to test this PR in a self-deployed VM before it is merged to develop.

hyi commented on hs_core/models.py:
```diff
@@ -1750,70 +1766,494 @@ class Meta:
+        """
+        if self.exists:
+            if self.fed_resource_file:
+                self.fed_resource_file.delete()
```
@alvacouch I see what you are doing now, and it looks like using fed_resource_file can cover the third scenario I mentioned above. However, we will need to make sure fed_resource_file to contain the full absolute iRODS path including irods zone in order for federated irods storage to delete the file properly.

alvacouch replied:
Migrations handle that already.

hyi commented on hs_core/models.py:
```diff
@@ -2025,26 +2523,20 @@ def get_crossref_deposit_xml(self, pretty_print=True):
-            pass
-
-        return total_file_size
+            return -1
```
@alvacouch returning -1 when there is an exception may break user tracking code when federation is not available. They want to continue counting where there is an exception for some files which may also have some utility for other use cases, but I can see the value to just return -1 rather than giving false impression that the returned size is valid even though some files in the resource have some issues with returning the correct size.

alvacouch replied:
What should I do in the case where I can't read the size? I only have three options: 
1. Raise an exception. 
2. Return a flag value
3. Return zero. 
Of these, only (1) makes any real sense.

alvacouch replied:
Modified to raise SessionException if iRODS fails.

hyi commented on hs_core/views/utils.py:
```diff
@@ -690,6 +657,7 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original):
     istorage.session.run("ibun", None, '-xDzip', zip_with_full_path, unzip_path)
     link_irods_folder_to_django(resource, istorage, unzip_path, (zip_fname,))
 
+    # TODO: why was the zipfile part of the django ResourceFile's?
```
@alvacouch Addressing you TODO comment, the zip file is part of the django ResourceFile since it is part of the resource to be unzipped. We provide calling function a option whether to remove the original zip file from the resource after it is unzipped.

alvacouch replied:
TODO removed. Thanks for explanation.

hyi commented on hs_geographic_feature_resource/receivers.py:
```diff
@@ -642,51 +651,60 @@ def geofeature_post_add_files_to_resource_handler(sender, **kwargs):
+        # from a federated # user account. In that case, it must be parsed after inclusion.
+        # Thus this script looks for that file in federated space only. Is this correct?
+        # If it is, then I would propose moving all of this to the pre_add script and doing
+        # the download there, for cohesion.
```
@zhiyuli Can you answer @alvacouch 's question above?

hyi commented:
@alvacouch I have finished code review for this big PR and left some comments. It looks good and thanks for taking on this refactoring to make resource file handling logic more elegant and simpler. I'd prefer to test it manually on a VM before merging, but if the federation test environment is not ready on a deploy VM, I am OK to have it merged first, then test it out on dev, and subsequently on beta before release deployment and patch things up if the manual tests reveal some regression issues.

alvacouch commented:
@pkdash @hyi One reason for the delay is that I needed to learn -- the hard way -- that exceptions raised in receivers do **not** become exceptions in the application! Darn. I was seeing an error in the geographic feature resource that was actually due to an exception that the application never saw.

alvacouch commented:
@hyi @pkdash: I have finished addressing all review comments and am engaged in testing that I didn't break anything. 

@mjstealey: @pkdash and I have agreed to the following merge schedule. 
1. I will cease work on this PR while he finishes the netCDF file format. 
2. When his PR is merged, I will merge and port his PR into this one. 
3. Then this one will get merged. This is so @pkdash doesn't have to do unnecessary work trying to keep up with me as I change this PR. 

Some caveats on this PR: 
1. calling BaseResource.size or ResourceFile.size can result in raising a SessionError if iRODS is not available. This is not accounted for in the UI.

pkdash commented:
@alvacouch 

> calling BaseResource.size or ResourceFile.size can result in raising a SessionError if iRODS is not available. This is not accounted for in the UI

What's the scenario when iRODS won't be available? How does the UI accessing the size information now which not likely to raise the same exception?

alvacouch commented:
@pkdash The current size implementation is stored rather than dynamically queried. I switched this to dynamically queried for a few reasons, the most pervasive of which was that it allows immediate detection of out-of-band changes to a file. Mostly these would be bugs. I could switch it back to stored and there will be no exception. However, there will also be no detection of cases in which a file was updated outside the ResourceFile API either. 

The cost of dynamic querying is that if iRODS suffers an outage -- however short -- this can force a "connection refused" for one of the requests for size, and then the whole size query for the resource fails with a SessionError.

alvacouch commented:
@pkdash Note also that the whole notion of size is a bit buggy when we don't have logical files, because at present, the geographic feature resource actually contains several physical files that are not referenced by a ResourceFile. Se note above on that. 

I really want your notion of logical files for the geographic feature resource!

pkdash commented:
@alvacouch 

> the geographic feature resource actually contains several physical files that are not referenced by a ResourceFile

I am not sure how it is possible. Looking at the code for feature resource it seems for each of those files there is an instance of ResourceFile.

alvacouch commented:
@pkdash I will check again. I was sure I found an instance where a file that was physical was implied rather than being an instance of a ResourceFile.... my evidence for this was that the ResourceFiles were not queried for the file; its name was computed from a different ResourceFile.

alvacouch commented:
@hyi @pkdash Pending final checks this is ready for merge. It is unreasonable to think that there will even be a one-day delay when a merge conflict does not emerge with this PR. 

Can we proceed to test it on an instance? 

@mjstealey This needs a test deploy.

alvacouch commented:
@pkdash @hyi @mjstealey Sorry, closed this without merging by mistake. Just reopened. 
Do I need to do something to kick off a build test?  

Sane comment applies: need a test build.

pkdash commented:
@alvacouch Looks like Jenkins is testing/building now.

alvacouch commented:
@mjstealey @pkdash @hyi This branch now passes all tests and is ready for beta deployment (despite the failure of the local testing environment). Michael, could you bring up a test server containing this?

alvacouch commented:
For unknown reasons, this PR breaks the use of `irods/use-local-irods.sh`. The iRODS containers exit prematurely for unknown reasons. See the script `irods/test-fed-irods.sh`. This script works with the develop branch, but fails for this branch. **Thus merging this branch will break federated testing for everyone.**

pkdash commented:
@alvacouch This PR has been deployed on riverdev.hydroshare.org.

pkdash commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -80,18 +83,17 @@ def create_bag_files(resource):
     # to_file_name = '{res_id}/data/visualization/'.format(res_id=resource.short_id)
```
Would be good to delete these commented code.

pkdash commented on hs_core/tests/api/rest/test_move_or_rename_file_or_folder.py:
```diff
@@ -53,27 +53,27 @@ def test_bad_requests(self):
         response = self.client.post(unzip_url, {
             "source_path": " ",
-            "target_path": "foo"
+            "target_path": "data/contents/foo"
```
Wondering why this change. Why not the hander for this endpoint do this fixed prefixing of directory path?

alvacouch replied:
@pkdash I think it should, but it doesn't, and I am unwilling to change it in this PR. Pre-existing condition. Reason: user might want to create bag-like structures outside data/contents. This test was writing outside data/contents before!

I am willing to coordinate fixing this with you. But I'd rather do it in a later PR. 

For now I have put guard clauses in the script so that it can't write outside data/contents. It shouldn't have to. 

This is just one example of the many kinds of potential file corruption I have found.

pkdash commented:
@alvacouch I have left couple of comments if you want to address. +1 from me.

hyi commented on hs_composite_resource/models.py:
```diff
@@ -147,13 +160,23 @@ def supports_delete_folder_on_zip(self, original_folder):
+                resource_file__startswith=full_path).all()
+        print("res_file_objects = ")
+        for f in res_file_objects:
+            print("   path={}".format(f.storage_path))
```
@alvacouch These print() debug statements need to be deleted.

hyi commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -11,50 +11,42 @@
     Parameters:
     :param resource: the resource to delete the bag for.
     :return: none
     """
```
@alvacouch It is better to update the comment as well to explicitly indicate ```delete the whole resource including files and bag```

alvacouch replied:
Modified docstring.

hyi commented:
@alvacouch I have left a couple of comments. Otherwise, +1 from me.

pkdash commented:
@alvacouch There are few review comments if you want to address. Once you resolve the conflicts and Jenkins checks are successful, you can merge. +1 from me.

alvacouch commented:
@pkdash @hyi @mjstealey I think I addressed all remaining review comments. Now rebasing. Dev environment still breaks. Phuong and I are working on it. The build of the irods containers fails and exits. I can't merge until we get to the bottom of that.

alvacouch commented:
@hyi: @pkdash raised an important issue with the REST endpoints: the zip and unzip take full pathnames data/contents/{folder}/file.zip rather than {folder}/file.zip. I understand why you would want this, because "it would be convenient" to be able to zip and unzip **outside** the bag space. However, it doesn't seem to be used anywhere and it gives users the ability to mistakenly corrupt the bag space (which tests for unzip actually did until I corrected them!).

Would it be safe to remove this option in a future PR and force all REST endpoints to act on things inside data/contents instead? Or is there some way this is used that I just can't find in the code?

hyi commented:
@alvacouch @pkdash I wrote these functions trying to be generally applicable without restricting input data path and assume these should be enforced by the calling routines. I think implementing this way has its merits to accommodate potential future changes to the path even if it does not look likely at this time. These zip/unzip utility functions are called by REST end point wrapper functions. Can REST end point wrapper functions enforce '/data/contents/' path rather than enforcing it on these utility function level?  I think something like if the input_path starts with /data/contents, then it can be passed in directly, and if it does not start with /data/contents, then /data/contents prefix can be prepended before calling zip/unzip utility routines to make sure all zipping/unzipping happens within /data/contents subfolders?

alvacouch commented:
@hyi That is what I am proposing: prepend data/contents at the REST level. Is there anything that can go wrong with **that?**

hyi commented:
@alvacouch That should work. I don't foresee any problem with that.

pkdash commented:
@alvacouch and @hyi It seems at least this REST endpoint (https://github.com/hydroshare/hydroshare/blob/1657-ResourceFile-API/hs_core/views/resource_folder_rest_api.py#L49) doing the prepend. 

I agree REST endpoint should do the prepend.

alvacouch commented:
@pkdash @hyi I want to do a bit more validation but I think I've merged the changes for @pkdash's changes correctly. I plan to merge over the weekend. I have family around so I think it'll likely get merged on Sunday afternoon after they leave. 

Alas, @pkdash, your new code for fixing the problems with composite resource would have been fixed by my PR. I enabled 'move-in-place' functionality so the logical file won't be touched. 

I am going to defer the fix for REST until the next PR, which will focus upon making all of REST and API consistent in their handling of data/contents. It fits better there. I've changed my mind on how to do that based upon the discussion above, and PR #1979 is obsolete. I will delete it and start over!

alvacouch commented:
@pkdash @hyi @mjstealey I have managed to consistently build iRODS tests via the following method: In hydroshare/irods, 

a.  run `./use-local-irods.sh` once to build the containers. This falls down. We'll fix this in a later step.
b.  run `./turn-off-local-irods.sh`, which has been modified to leave the iRODS containers alone but resets everything else. 
c.  run `./use-local-irods.sh` again with the containers already up. 

This avoids the download step for building the irods containers and -- for some reason -- consistently succeeds. 

This is a bit contorted but local irods testing is again possible using this technique. 

Thus, I am confident that the problem with testing is not the content of the PR.
